### PR TITLE
SSN and Phone country code fixes

### PR DESCRIPTION
**Pull request by guymaness2**
##### What does the change add or fix?

* Fixed bug causing SSN type attributes to not be encrypted properly on import.  
* Added CountryCode to phone number csv import and fixed logic to better deal with international phone numbers.  
* Added logic to create Country Code defined values in Rock as needed. Current support is limited to Australia, Cambodia, Germany, Ghana, Niger and Spain. Plans are to increase supported countries as needed by KFS driven migrations.

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

* Fixed bug causing SSN type attributes to not be encrypted properly on import.  
* Added CountryCode to PhoneNumber csv import and fixed logic to better deal with international phone numbers.  
* New logic to add "Country Code" defined values in Rock as country codes are encountered. Current support is limited to Australia, Cambodia, Germany, Ghana, Niger and Spain. Plans are to increase supported countries as needed by KFS driven migrations.

---------

##### Who reported, requested, or paid for the change?

Warranty

---------

##### Does this update or add options to the block UI?

Person Detail phone number country code dropdown:  
image  
  
New Country Code defined values:  
image

---------

##### What files does it affect?

* Bulldozer.CSV/CSVComponent.cs
* Bulldozer.CSV/Maps/Phone.cs
* Bulldozer.CSV/SampleCSVs/01 - Individual.csv
* Bulldozer.CSV/SampleCSVs/22 - PhoneNumber.csv
* Bulldozer/Utility/AddMethods.cs  

---------

##### Is it a breaking change for other versions/clients?

Requires new CountryCode column to be present in PhoneNumber.csv import file.

### CODE REVIEW

**Comment by guymaness2** [comment]
@nateh777 ready for review

File: Bulldozer.CSV/CSVPhoneCountryCode.cs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            CountryCode = 49,
+            Description = "German 4 digit prefix Mobile Phone Number with Country Code",
+            MatchExpression = @"^49(\d{4})(\d{7})$",
+            FormatExpression = @"$1 $2"
+        };
+
+        public static CountryCodeData Ghana = new CountryCodeData
+        {
+            CountryCode = 233,
+            Description = "Ghanan Phone Number",
```
Comment (guymaness2) [inline Bulldozer.CSV/CSVPhoneCountryCode.cs:95]:
```suggestion
            Description = "Ghanaian Phone Number",
```

File: Bulldozer.CSV/CSVPhoneCountryCode.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            MatchExpression = @"^(\d{2})(\d{3})(\d{4})$",
+            FormatExpression = @"$1 $2 $3"
+        };
+
+        public static CountryCodeData GhanaWithCountry = new CountryCodeData
+        {
+            CountryCode = 233,
+            Description = "Ghanan Phone Number with Country Code",
```
Comment (guymaness2) [inline Bulldozer.CSV/CSVPhoneCountryCode.cs:103]:
```suggestion
            Description = "Ghanaian Phone Number with Country Code",
```

File: Bulldozer.CSV/Maps/Phone.cs [hunk 3]
```diff
@@ -172,6 +172,71 @@ private static void SavePhoneNumbers( List<PhoneNumber> phoneNumberList )
                 rockContext.SaveChanges( DisableAuditing );
             } );
         }
+
+        /// <summary>
+        /// Check for existing CountryCode defined values or create new.
+        /// </summary>
+        /// <param name="phoneNumberList">The list of phone numbers.</param>
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Phone.cs:179]:
```suggestion
        /// <param name="countryCode">The country code.</param>
```

File: Bulldozer.CSV/Maps/Phone.cs [hunk 4]
```diff
@@ -172,6 +172,71 @@ private static void SavePhoneNumbers( List<PhoneNumber> phoneNumberList )
...
+                            AttributeId = ccMatchExprAttribute.Id,
+                            Value = ccData.MatchExpression
+                        } );
+                        var ccMatchExprAttributeVal = new AttributeValue
+                        {
+                            EntityId = countryCodeDVCache.Id,
+                            AttributeId = ccMatchExprAttribute.Id,
+                            Value = ccData.MatchExpression
+                        };
+                        attributeValueService.Add( ccMatchExprAttributeVal );
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Phone.cs:216]:
```suggestion
                        var ccMatchExprAttributeVal = new AttributeValue
                        {
                            EntityId = countryCodeDVCache.Id,
                            AttributeId = ccMatchExprAttribute.Id,
                            Value = ccData.MatchExpression
                        };
                        attributeValueService.Add( ccMatchExprAttributeVal );
                        
                        countryCodeDVCache.AttributeValues.Remove( ccMatchExprAttribute.Key );
                        countryCodeDVCache.AttributeValues.Add( ccMatchExprAttribute.Key, new AttributeValueCache
                        {
                            AttributeId = ccMatchExprAttribute.Id,
                            Value = ccData.MatchExpression
                        } );
```

File: Bulldozer.CSV/Maps/Phone.cs [hunk 5]
```diff
@@ -172,6 +172,71 @@ private static void SavePhoneNumbers( List<PhoneNumber> phoneNumberList )
...
+                            AttributeId = ccFormatExprAttribute.Id,
+                            Value = ccData.FormatExpression
+                        } );
+                        var ccFormatExprAttributeVal = new AttributeValue
+                        {
+                            EntityId = countryCodeDVCache.Id,
+                            AttributeId = ccFormatExprAttribute.Id,
+                            Value = ccData.FormatExpression
+                        };
+                        attributeValueService.Add( ccFormatExprAttributeVal );
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Phone.cs:234]:
```suggestion
                        var ccFormatExprAttributeVal = new AttributeValue
                        {
                            EntityId = countryCodeDVCache.Id,
                            AttributeId = ccFormatExprAttribute.Id,
                            Value = ccData.FormatExpression
                        };
                        attributeValueService.Add( ccFormatExprAttributeVal );
                        
                        countryCodeDVCache.AttributeValues.Remove( ccFormatExprAttribute.Key );
                        countryCodeDVCache.AttributeValues.Add( ccFormatExprAttribute.Key, new AttributeValueCache
                        {
                            AttributeId = ccFormatExprAttribute.Id,
                            Value = ccData.FormatExpression
                        } );
```

File: Bulldozer.CSV/SampleCSVs/22 - PhoneNumber.csv [hunk 6]
```diff
@@ -1,3 +1,4 @@
-PhonePersonId,PhoneType,Phone,PhoneIsMessagingEnabled,PhoneIsUnlisted,PhoneId
-25570361,Fax,(123) 456-7890,0,0,25570361_1000
-142,Checkin Number,(864) 555-5555,0,1,142_1003
+PhonePersonId,PhoneType,Phone,PhoneIsMessagingEnabled,PhoneIsUnlisted,PhoneId,CountryCode
+25570361,Fax,(123) 456-7890,0,0,25570361_1000,
+142,Checkin Number,(864) 555-5555,0,1,142_1003,
+182,Home,2 3456 7699,0,0,182_265,61
```
Comment (guymaness2) [inline Bulldozer.CSV/SampleCSVs/22 - PhoneNumber.csv:4]:
```suggestion
182,Home,234 567 890,0,0,182_265,61
```

File: Bulldozer.CSV/SampleCSVs/01 - Individual.csv [hunk 7]
```diff
@@ -6,3 +6,4 @@ FamilyId,FamilyName,CreatedDate,PersonId,Prefix,FirstName,NickName,MiddleName,La
 968111412207101120A,Meese Family,,146,Mrs,Sara,,,Meese,,Adult,Married,Attendee,Active,,,864-938-1593,,No,smeese@gmail.com,Yes,,Female,6/9/1985,,,,,,,,"workingmom@gmail.com,smeese@fbc.org",10/11/2008,blah1,blah2,blah3
 968111412207101120A,Meese Family,,147,,Oliver,,,Meese,,Child,,,,,,,,,,,,Male,5/1/2011,,,,,,,"5,981,853",,,,,
 968111412207101120A,Meese Family,,148,,Debra,,,Meese,,Child,,,,,,,,,,,,Female,1/1/2009,,,,,,,,,,,,
+968111412207101126A,Decker Family,,182,Mr,Ted,,,Decker,,Adult,,,,,,,,,,,,Male,,,,,,,,,,,,,
```
Comment (guymaness2) [inline Bulldozer.CSV/SampleCSVs/01 - Individual.csv:9]:
```suggestion
968111412207101126A,Decker Family,,182,Mr,Michael,Mick,,Dundee,,Adult,,,,,,,,,,,,Male,,,,,,,,,,,,,
```

**Review by nateh777 — APPROVED** [review]

**Pull request merged**